### PR TITLE
Add differential abundance

### PR DESCRIPTION
**Pull request by asizemore**
Relies on https://github.com/VEuPathDB/Rserve/pull/17 and https://github.com/VEuPathDB/EdaDataService/pull/267

Adds a plugin for the differential abundance compute. Given samples tested for the abundance of genes or taxa, differential abundance finds those genes or taxa that are more present in one group of samples as compared to the second group, and vv.

### CODE REVIEW

**Comment by asizemore** [comment]
Note there is some code left in there for making fake absolute data from relative abundance data. 

Also i have stubbed in `groupA` and `groupB` until we add `listToRVector` into eda common (i'll start on that tomorrow). But the below should work with this branch

```
{
	"config": {
		"collectionVariable": {
			"variableId": "EUPATH_0009257",
			"entityId": "EUPATH_0000813"
		},
		"comparisonVariable": {
			"variableId": "PATO_0000047",
			"entityId": "EUPATH_0000096"
		},
		"differentialAbundanceMethod": "DESeq",
		"differentialAbundanceGroupA": ["groupA"],
		"differentialAbundanceGroupB": ["groupB"]
	},
	"derivedVariables": [],
	"filters": [],
	"studyId": "EMP_1064_Bee_Microbiome-1"
}
```

File: schema/url/computes/differentialAbundance.raml [hunk 1]
```diff
@@ -0,0 +1,20 @@
...
+
+  DifferentialAbundancePluginRequest:
+    type: ComputeRequestBase
+    properties:
+      config: DifferentialAbundanceComputeConfig
+
+  DifferentialAbundanceComputeConfig:
+    properties:
+      collectionVariable: VariableSpec
+      comparisonVariable: VariableSpec
```
Comment (d-callan) [inline schema/url/computes/differentialAbundance.raml]: These three are clearly related to one another. maybe a new raml type? 

ComparatorSpec:
  properties:
    variable:
    groupA:
    groupB:

File: schema/url/computes/differentialAbundance.raml [hunk 2]
```diff
@@ -0,0 +1,20 @@
...
+      differentialAbundanceGroupA: string[]
+      differentialAbundanceGroupB: string[]
+      differentialAbundanceMethod: DifferentialAbundanceMethod
```
Comment (d-callan) [inline schema/url/computes/differentialAbundance.raml:14]: im not convinced we need to repeat differentialAbundance so many times here..

File: src/main/java/org/veupathdb/service/eda/compute/plugins/differentialabundance/DifferentialAbundancePlugin.java [hunk 3]
```diff
@@ -0,0 +1,130 @@
...
+
+      // Read in the sample metadata
+      List<VariableSpec> sampleMetadataVars = ListBuilder.asList(comparisonVariableSpec);
+      sampleMetadataVars.add(computeEntityIdVarSpec);
+      System.out.println(sampleMetadataVars);
+      connection.voidEval(util.getVoidEvalFreadCommand(INPUT_DATA, sampleMetadataVars));
+      connection.voidEval("sampleMetadata <- " + INPUT_DATA);
+
+
+      // TODO make a helper for this i think
```
Comment (d-callan) [inline src/main/java/org/veupathdb/service/eda/compute/plugins/differentialabundance/DifferentialAbundancePlugin.java]: if it makes sense to put this in edacommon, then you should probably do that before releasing a new version.

File: src/main/java/org/veupathdb/service/eda/compute/plugins/differentialabundance/DifferentialAbundancePlugin.java [hunk 4]
```diff
@@ -0,0 +1,130 @@
...
+        if (first) {
+          first = false;
+          dotNotatedIdColumnsString = dotNotatedIdColumnsString + util.singleQuote(idCol);
+        } else {
+          dotNotatedIdColumnsString = dotNotatedIdColumnsString + "," + util.singleQuote(idCol);
+        }
+      }
+      dotNotatedIdColumnsString = dotNotatedIdColumnsString + ")";
+
+      // TEMP FOR TESTING ONLY - REMOVE WHEN ABSOLUTE ABUNDANCES ARE HERE
```
Comment (d-callan) [inline src/main/java/org/veupathdb/service/eda/compute/plugins/differentialabundance/DifferentialAbundancePlugin.java:98]: how is that going btw? getting absolute abundances loaded and collections annotated and constrained?

File: src/main/java/org/veupathdb/service/eda/compute/plugins/differentialabundance/DifferentialAbundancePlugin.java [hunk 5]
```diff
@@ -0,0 +1,130 @@
...
+      connection.voidEval("taxaColNames <- names(absoluteAbundanceData[, -c('" + computeEntityIdColName + "', as.character(" + dotNotatedIdColumnsString + "))])");
+      connection.voidEval("absoluteAbundanceData[, (taxaColNames) := lapply(.SD,function(x) {round(x*1000)}), .SDcols=taxaColNames]");
+      // END OF TEMP FOR TESTING
+
+      connection.voidEval("diffabundDT <- microbiomeComputations::AbsoluteAbundanceData(data=absoluteAbundanceData" +
```
Comment (d-callan) [inline src/main/java/org/veupathdb/service/eda/compute/plugins/differentialabundance/DifferentialAbundancePlugin.java]: these are pretty awfully named...

File: src/main/java/org/veupathdb/service/eda/compute/plugins/differentialabundance/DifferentialAbundancePlugin.java [hunk 6]
```diff
@@ -0,0 +1,130 @@
...
+                                                          ", comparisonVariable=" + PluginUtil.singleQuote(comparisonVariable) +
+                                                          ", groupA=" + groupA +
+                                                          ", groupB=" + groupB + 
+                                                          ", method=" + PluginUtil.singleQuote(method) + 
+                                                          ", verbose=TRUE)");
+
+
+
+      String dataCmd = "writeData(differentialabundanceDT, NULL, TRUE)";
+      String metaCmd = "writeMeta(differentialabundanceDT, NULL, TRUE)";
```
Comment (d-callan) [inline src/main/java/org/veupathdb/service/eda/compute/plugins/differentialabundance/DifferentialAbundancePlugin.java]: if we dont mean to use the data and meta endpoints, do we need to write them? im not sure..

**Comment by d-callan** [comment]
doesnt necessarily need to be this pr i suppose, but we should remember that this plugin is going to have to set some constraints against different types of collections. and we need to decide how to handle that for other plugins as well. meaning, do we introduce constraints against absolute abundances there?

File: src/main/java/org/veupathdb/service/eda/compute/plugins/differentialabundance/DifferentialAbundancePlugin.java (hunk 4 above)
Comment (asizemore) [reply src/main/java/org/veupathdb/service/eda/compute/plugins/differentialabundance/DifferentialAbundancePlugin.java:98]: lol bad. I keep poking people and getting radio silence.

File: src/main/java/org/veupathdb/service/eda/compute/plugins/differentialabundance/DifferentialAbundancePlugin.java (hunk 6 above)
Comment (ryanrdoherty) [reply src/main/java/org/veupathdb/service/eda/compute/plugins/differentialabundance/DifferentialAbundancePlugin.java]: No, you do not! :)  If you don't, then requests to those file endpoints should return 404, which I think is what we want.

**Comment by asizemore** [comment]
> do we introduce constraints against absolute abundances there?

I guess we'll see if any movement happens on that front. I'll bother folks again tomorrow at the mbio meeting. If it looks like that's going to happen pretty soon, I'd be happy to try to get constraints into this PR. Otherwise i'll move forward with merging just the plugin

**Comment by asizemore** [comment]
I'm still working on getting the app constraints meeting to happen, so i think that should be a different PR

**Review by ryanrdoherty — APPROVED** [review]

**Pull request merged**